src/manifest/__tests__/manifest.test.ts

```diff
@@ -0,0 +1,721 @@
+    const Manifest = require("../manifest").default;
+    const manifest = new Manifest(oldManifestArgs, {});
+
+    const eeSpy = jest.spyOn(manifest, "trigger").mockImplementation(jest.fn());
+
+    const [oldPeriod1, oldPeriod2] = manifest.periods;
+
+    const newAdaptations = {};
+    const newPeriod1 = { id: "foo0" };
+    const newPeriod2 = { id: "foo1" };
+    const newManifest = {
+        _duration: 13,
+      adaptations: newAdaptations,
+      availabilityStartTime: 6,
+      baseURL: "test2",
+      getDuration() { return 13; },
+      id: "man2",
+      isLive: true,
+      lifetime: 14,
+      minimumTime: 5,
+      parsingErrors: ["c", "d"],
+      presentationLiveGap: 19,
+      suggestedPresentationDelay: 100,
+      timeShiftBufferDepth: 3,
+      periods: [newPeriod1, newPeriod2],
+      transport: "foob",
+      uris: ["url3", "url4"],
+    };
+
+    manifest.update(newManifest);
+    expect(manifest.adaptations).toEqual(newAdaptations);
+    expect(manifest.availabilityStartTime).toEqual(6);
+    expect(manifest.baseURL).toEqual("test2");
+    expect(manifest.getDuration()).toEqual(13);
+    expect(manifest.id).toEqual("man2");
+    expect(manifest.isLive).toEqual(true);
+    expect(manifest.lifetime).toEqual(14);
+    expect(manifest.minimumTime).toEqual(5);
+    expect(manifest.parsingErrors).toEqual(["c", "d"]);
+    expect(manifest.presentationLiveGap).toEqual(19);
+    expect(manifest.suggestedPresentationDelay).toEqual(100);
+    expect(manifest.timeShiftBufferDepth).toEqual(3);
+    expect(manifest.transport).toEqual("foob");
+    expect(manifest.uris).toEqual(["url3", "url4"]);
+
+    expect(manifest.periods).toEqual([newPeriod1, newPeriod2]);
+
+    expect(fakeUpdatePeriodInPlace).toHaveBeenCalledTimes(2);
+    expect(fakeUpdatePeriodInPlace).toHaveBeenCalledWith(oldPeriod1, newPeriod1);
+    expect(fakeUpdatePeriodInPlace).toHaveBeenCalledWith(oldPeriod2, newPeriod2);
+    expect(eeSpy).toHaveBeenCalledTimes(1);
+    expect(eeSpy).toHaveBeenCalledWith("manifestUpdate", null);
+    expect(logSpy).not.toHaveBeenCalled();
+    logSpy.mockRestore();
+    eeSpy.mockRestore();
+  });
+
+  it("should prepend older Periods when calling `update`", () => {
+    const oldManifestArgs = {
+      availabilityStartTime: 5,
+      baseURL: "test",
+      duration: 12,
+      id: "man",
+      isLive: false,
+      lifetime: 13,
+      minimumTime: 4,
+      parsingErrors: ["a", "b"],
+      periods: [{ id: "1" }],
+      presentationLiveGap: 18,
+      suggestedPresentationDelay: 99,
+      timeShiftBufferDepth: 2,
+      transportType: "foobar",
+      uris: ["url1", "url2"],
+    };
+
+    const logSpy = jest.spyOn(log, "info").mockImplementation(jest.fn());
+    const fakePeriod = jest.fn((period) => {
+      return { ...period, id: "foo" + period.id, parsingErrors: [period.id] };
+    });
+    const fakeUpdatePeriodInPlace = jest.fn((oldPeriod, newPeriod) => {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Update or refactor this function so that its implementation doesn't duplicate the one on line 360. ([typescript:S4144](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4144))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=440&issues=AWgpL7_oUMkE2J58feAg&open=AWgpL7_oUMkE2J58feAg)

[ISSUE_KEY]: AWgpL7_oUMkE2J58feAg